File: Hard/2025-10-17-3003-Maximize-the-Number-of-Partitions-After-Operations/solution.js
var maximizePartitions = function(s, k) {
    const getMaxPartitions = (str, k) => {
        let count = 0;
        let left = 0;
        const charCount = new Map();

        for (let right = 0; right < str.length; right++) {
            charCount.set(str[right], (charCount.get(str[right]) || 0) + 1);

            while (charCount.size > k) {
                charCount.set(str[left], charCount.get(str[left]) - 1);
                if (charCount.get(str[left]) === 0) {
                    charCount.delete(str[left]);
                }
                left++;
            }
            count++;
        }
        return count;
    };

    const originalPartitions = getMaxPartitions(s, k);
    let maxPartitions = originalPartitions;

    for (let i = 0; i < s.length; i++) {
        const originalChar = s[i];
        for (let j = 0; j < 26; j++) {
            const newChar = String.fromCharCode('a'.charCodeAt(0) + j);
            if (newChar !== originalChar) {
                const modifiedString = s.slice(0, i) + newChar + s.slice(i + 1);
                maxPartitions = Math.max(maxPartitions, getMaxPartitions(modifiedString, k));
            }
        }
    }

    return maxPartitions;
};

File: Hard/2025-10-17-3003-Maximize-the-Number-of-Partitions-After-Operations/solution.py
class Solution:
    def maximizePartitions(self, s: str, k: int) -> int:
        def count_partitions(s: str, k: int) -> int:
            count = 0
            left = 0
            char_count = {}
            for right in range(len(s)):
                char_count[s[right]] = char_count.get(s[right], 0) + 1
                while len(char_count) > k:
                    char_count[s[left]] -= 1
                    if char_count[s[left]] == 0:
                        del char_count[s[left]]
                    left += 1
                count += 1
            return count

        max_partitions = count_partitions(s, k)
        for i in range(len(s)):
            original_char = s[i]
            for new_char in 'abcdefghijklmnopqrstuvwxyz':
                if new_char != original_char:
                    modified_s = s[:i] + new_char + s[i+1:]
                    max_partitions = max(max_partitions, count_partitions(modified_s, k))
        return max_partitions